crypto/ec/ecp_nistp256.c

```diff
@@ -1875,10 +1875,13 @@ static NISTP256_PRE_COMP *nistp256_pre_comp_new(void)
 
 NISTP256_PRE_COMP *EC_nistp256_pre_comp_dup(NISTP256_PRE_COMP *p)
 {
-    int i;
-    if (p != NULL)
-        CRYPTO_UP_REF(&p->references, &i, p->lock);
-    return p;
+    int ref = 0;
+
+    if (p != NULL) {
+        if (!CRYPTO_UP_REF(&p->references, &ref, p->lock))
+            return NULL;
+    }
+    return ref > 0 ? p : NULL;
```

Comment:
Hmmm, thought experiment:

```
    int ref = 0;

    if (p == NULL
        || !CRYPTO_UP_REF(&p->references, &ref, p->lock)
        || ref <= 0)
        return NULL;
    return p;
```

I kinda find the single returns clearer.

Also, the check of `ref` on an UP_REF seems curious.  We *do* need to make such checks on DOWN_REF to see when further action is needed, but on UP_REF?

Reply:
The check for ref <= 0 is an overflow check, although I‘m not sure whether it is really necessary in practice.

Reply:
If we want to include an overflow check, it should be part of `CRYPTO_UP_REF` function.  That way it immediately goes everywhere.

Personally, I don't see a need for this but wouldn't object to it being put in.

Reply:
I looked at doing that in CRYPTO_UP_REF, but it won't work becase
that is defined in tons of places, most of the time it is just a macro
calling CRYPTO_atomic_add, and while reference counts are never negative,
an atomic int is not a reference, and can therefore also be negative.
So I think the best we can do is like X509_up_ref:
```C
int X509_up_ref(X509 *x)
{
    int i;

    if (CRYPTO_UP_REF(&x->references, &i, x->lock) <= 0)
        return 0;

    REF_PRINT_COUNT("X509", x);
    REF_ASSERT_ISNT(i < 2);
    return ((i > 1) ? 1 : 0);
}
```